core/src/worker/workflow/machines/upsert_search_attributes_state_machine.rs

```diff
@@ -238,12 +355,94 @@ mod tests {
         .try_into()
         .unwrap();
 
-        OnEventWrapper::on_event_mut(&mut sm, cmd_scheduled_sm_event)
-            .expect("CommandScheduled should transition Created -> CommandIssued");
+        OnEventWrapper::on_event_mut(
+            &mut sm,
+            CommandType::UpsertWorkflowSearchAttributes
+                .try_into()
+                .unwrap(),
+        )
+        .expect("CommandScheduled should transition Created -> CommandIssued");
         assert_eq!(CommandIssued {}.to_string(), sm.state().to_string());
 
-        OnEventWrapper::on_event_mut(&mut sm, cmd_recorded_sm_event)
-            .expect("CommandRecorded should transition CommandIssued -> Done");
-        assert_eq!(Done {}.to_string(), sm.state().to_string());
+        let recorded_res = OnEventWrapper::on_event_mut(&mut sm, cmd_recorded_sm_event);
+        if nondetermistic {
+            assert_matches!(
+                recorded_res.unwrap_err(),
+                MachineError::Underlying(WFMachinesError::Nondeterminism(_))
+            );
+        } else {
+            recorded_res.expect("CommandRecorded should transition CommandIssued -> Done");
+            assert_eq!(Done {}.to_string(), sm.state().to_string());
+        }
+    }
+
+    #[rstest::rstest]
+    #[tokio::test]
+    async fn upserting_change_version_directly(
+        #[values(true, false)] flag_in_history: bool,
+        #[values(true, false)] with_patched_cmd: bool,
+    ) {
+        let patch_id = "whatever".to_string();
+        let mut t = TestHistoryBuilder::default();
+        t.add_by_type(EventType::WorkflowExecutionStarted);
+        t.add_full_wf_task();
+        if flag_in_history {
+            t.set_flags_first_wft(
+                &[CoreInternalFlags::UpsertSearchAttributeOnPatch as u32],
+                &[],
+            );
+        }
+        if with_patched_cmd {
+            t.add_has_change_marker(&patch_id, false);
+        }
+        t.add_upsert_search_attrs_for_patch(&[patch_id.clone()]);
+        t.add_full_wf_task();
```

Comment:
nit: this history wouldn't be generated by a real workflow, there would need to be something that prevents the workflow from completing in the first task but that probably wouldn't change the outcome of the test.

Reply:
Not true. Waiting on a signal can result in a history which looks just like this

Reply:
But then a signal would be delivered to unblock the workflow

Reply:
Lol, yeah fair point. It doesn't matter for the test though indeed.